### PR TITLE
pgx mirage

### LINKED ISSUE #67: Mirage support
#57 and #58 are both attempts by @paurkedal to support Mirage. The point of this ticket is to:

1. Get Mirage support working
2. Ensure it stays working

For Mirage support, I'm wondering if this would be Pgx_unix, or another new Mirage-specific one?

Is there a way we can run a CircleCI build to ensure this works?

### PR DESCRIPTION
**Pull request by anuragsoni**
**TODO**

- [x] Update CI to build/test that the unikernel works
- [x] Add documentation for how to build/run the unikernel example

Closes #67

### CODE REVIEW

**Comment by anuragsoni** [comment]
Tested manually by running the example unikernel against a postgres instance running on google cloud.

File: pgx_lwt/src/pgx_lwt.ml [hunk 1]
```diff
@@ -1,75 +1,65 @@
...
-      let len = Array.length h_addr_list in
-      let i = Random.int len in
-      let addr = h_addr_list.(i) in
-      Unix.ADDR_INET (addr, port))
-    >>= Lwt_io.open_connection
-  ;;
+    let close_in = IO.close_in
+    let open_connection = IO.open_connection
+    let getlogin = IO.getlogin
+    let debug s = Logs_lwt.debug (fun m -> m "%s" s)
```
Comment (brendanlong) [inline pgx_lwt/src/pgx_lwt.ml:49]: It's sort of weird that we're assuming anyone using Lwt wants to use Logs, but I guess our logging just sucks in general.

Maybe one day we can make Pgx just require Logs in general..

File: pgx_lwt/src/s.ml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  val input_char : in_channel -> char Lwt.t
+  val really_input : in_channel -> bytes -> int -> int -> unit Lwt.t
+  val close_in : in_channel -> unit Lwt.t
+  val getlogin : unit -> string Lwt.t
+  val open_connection : sockaddr -> (in_channel * out_channel) Lwt.t
+end
+
+module type Pgx_impl = sig
+  include Pgx.S with type 'a IO.t = 'a Lwt.t
+end
```
Comment (brendanlong) [inline pgx_lwt/src/s.ml:21]: Why do we need this? Isn't this just `Pgx.S with type IO.t := Lwt.t`?

File: pgx_lwt_mirage/src/pgx_lwt_mirage.mli [hunk 3]
```diff
@@ -0,0 +1,28 @@
+(* Draft for cherry-picking into pgx.
```
Comment (brendanlong) [inline pgx_lwt_mirage/src/pgx_lwt_mirage.mli:1]: Weird comment?

File: pgx_test/src/dune [hunk 4]
```diff
@@ -1,3 +1,3 @@
 (library
  (name pgx_test)
- (libraries alcotest base64 pgx))
+ (libraries alcotest base64 pgx pgx_value_core))
```
Comment (brendanlong) [inline pgx_test/src/dune:3]: Why does Pgx_test need a dependency on this? I thought Pgx_value_core handled its own testing.

File: pgx_lwt/src/pgx_lwt.ml (hunk 1 above)
Comment (anuragsoni) [reply pgx_lwt/src/pgx_lwt.ml:49]: logs is the only logging option i know of that works with lwt so it felt like using it for the lwt backend was fine as its more likely to be pulled in by lwt users. If pgx itself ends up depending on logs then the `debug` implementation can be moved up to the core library as well.

File: pgx_lwt/src/s.ml (hunk 2 above)
Comment (anuragsoni) [reply pgx_lwt/src/s.ml:21]: Good catch. I had initially done this as i need to return a first class module from the pgx_lwt_mirage.connect function and parameterized types aren't supported for packed modules. But I didn't realize that I can just create a temporary alias inside the mirage backend an then i won't need to create this temporary instance. I'll update the PR

File: pgx_lwt_mirage/src/pgx_lwt_mirage.mli (hunk 3 above)
Comment (anuragsoni) [reply pgx_lwt_mirage/src/pgx_lwt_mirage.mli:1]: I think this came from when i copied the license header from the gist linked to in the issue for mirage. I wasn't sure what content i'm allowed to touch in such situation

File: pgx_test/src/dune (hunk 4 above)
Comment (anuragsoni) [reply pgx_test/src/dune:3]: this should be removed. I'll update

File: pgx_lwt_mirage/src/pgx_lwt_mirage.mli (hunk 3 above)
Comment (brendanlong) [reply pgx_lwt_mirage/src/pgx_lwt_mirage.mli:1]: We can edit this comment as long as we leave the license block alone and only add to the copyright section.

File: pgx_test/src/dune (hunk 4 above)
Comment (anuragsoni) [reply pgx_test/src/dune]: This is now removed

File: pgx_lwt_mirage/src/pgx_lwt_mirage.mli (hunk 3 above)
Comment (anuragsoni) [reply pgx_lwt_mirage/src/pgx_lwt_mirage.mli]: Updated the header to remove the comment about cherry-picking

File: pgx_lwt/src/s.ml (hunk 2 above)
Comment (anuragsoni) [reply pgx_lwt/src/s.ml]: The module alias is now just Pgx_lwt.S and is shared across the different lwt backends

File: pgx/src/pgx.ml [hunk 5]
```diff
@@ -471,7 +471,8 @@ module type IO = Io_intf.S
 module type S = Pgx_intf.S
 
 module Make (Thread : IO) = struct
-  open Thread
+  module IO = Thread
+  open IO
```
Comment (brendanlong) [inline pgx/src/pgx.ml:475]: What's the reason for this change? It looks like it creates a bunch of downstream complexity (like the `Pgx_lwt.Io_intf` module) but doesn't seem to change anything else?

Comment (brendanlong) [reply pgx/src/pgx.ml:475]: It looks like we did this to accomidate Pgx_lwt, but I think we should keep our interface here simple and let Pgx_lwt do more complicated things to generate the interface if necessary. I'll try to make an example.
